### DISCUSSION TITLE
How to set a variable?

### DISCUSSION THREAD

**Discussion by tobka777** [devonfw/IDEasy#291] [Q&A] [answered]
How to set a variable?

**Comment by tobka777** [comment]
Here is a link how to do it..

**Comment by hohwille** [comment, marked as answer]
First of all see the [available variables documentation](https://github.com/devonfw/IDEasy/blob/main/documentation/variables.adoc) so you know what you can configure.
Variables like `PATH` are configured globally and affect your entire system. The instructions how to set or modify them depends on your operating system:
* [set environment variable on Windows](https://phoenixnap.com/kb/windows-set-environment-variable#ftoc-heading-4)
* [set environment variable on MacOS](https://phoenixnap.com/kb/set-environment-variable-mac#ftoc-heading-5)
* [set environment variable on Linux](https://www.freecodecamp.org/news/how-to-set-an-environment-variable-in-linux/#how-to-make-environment-variables-persistent-in-linux)

Typically you do not want to set a variable globally but only for IDEasy.
Therefore you can edit a configuration file in the following places (as described in [configuration](https://github.com/devonfw/IDEasy/blob/main/documentation/configuration.adoc)):
* `$IDE_HOME/conf/` for configurations that only apply to you in the project identified by `$IDE_HOME`
* `$IDE_HOME/settings/` for configurations shared with the entire team of your project identified by `$IDE_HOME`. Be aware that changes made here need to be managed with git. For details also see [settings](https://github.com/devonfw/IDEasy/blob/main/documentation/settings.adoc).
* `~/.ide/` for personal configurations that apply to all your IDEasy projects.

NOTE: `$IDE_HOME` refers to the IDEasy project folder (e.g. `C:/projects/my-cool-project` on Windows or `~/projects/my-cool-project` on MacOS or Linux) and `~` refers to the users home directory. If you are a Windows user and do not know where your home directory is, then open Windows Explorer (Press `[Windows][r]`, enter `explorer` and hit `[Return]`) and then type in the location (e.g. `Quick access`) the following text `%USERPROFILE%` and press `[Return]`.

If you found the proper directory where to configure the variable, next you need to pick the correct configuration file:
* `ide.properties` is the correct choice if it exists
* `devon.properties` should be used if it exists but there is no `ide.properties`
* if none of these files exists, create the file `ide.properties`

Now use the editor of your choice to edit the configuration file (`ide.properties`):
* Search for the variable you want to set, if it is already set in your configuration file so you only need to change its value to what you want to have.
* Otherwise add a new line to the end of the file and set the variable in the form of `«VARIABLE»=«VALUE»`. Example:
```
M2_REPO=~/.m2/repository
```
* In case you also want to export the variable so it is available to other processes and tools (e.g. IntelliJ, Eclipse, maven, npm, etc.) prefix the variable definition with the `export` keyword. Example:
```
export http_proxy=http://proxy.customer.com
```
